ui/numbered_menu.py

```diff
@@ -28,22 +34,108 @@ def __init__(self, *args, **kwargs):
             "KEY_9": 9,
             "KEY_0": 0,
         }
+        self.entry_count = len(self.contents)
+        self.last_input_time = 0
+        self.current_input = None
 
     def activate(self):
         self.i.set_streaming(self.on_key_pressed)
-        Menu.activate(self)
+        # Menu.activate(self)
+        self.before_activate()
+        self.to_foreground()
+        if not self.is_multi_digit():
+            Menu.activate(self)  # no need to check for character state since we are in instant mode
+            return
+
+        self.last_input_time = -self.input_delay
+        while self.in_background:  # Most of the work is done in input callbacks
+            sleep(0.1)
+            self.idle_loop()
+            self.check_character_state()
 
     def deactivate(self):
         Menu.deactivate(self)
         self.i.remove_streaming()
 
     def on_key_pressed(self, key):
```

Comment:
You should definitely wrap that in `to_be_foreground`: If a NumberedMenu calls other UI element and goes into background, we can't continue processing callbacks from that NumberedMenu.